### PR TITLE
[TE] - Use client from config, to support external clients

### PR DESCRIPTION
**Pull request by npawar**
This pull request introduces clients.yml file, to define configs for all clients we will use in Thirdeye. The client to use will be defined in the individual Thirdeye requests.
Currently, all requests are going to the default Pinot client.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java [hunk 1]
```diff
@@ -2,17 +2,32 @@
...
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+
+  public ThirdEyeClient(String name) {
+    this.name = name;
+  }
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java]: If there's only one ThirdEye client instance per class, maybe go without a separate name property

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/pinot/PinotThirdEyeClientConfig.java [hunk 2]
```diff
@@ -94,12 +81,31 @@ public String toString() {
...
+        if (StringUtils.isNotBlank(brokerUrl)) {
+          config.setBrokerUrl(brokerUrl);
+        }
+        if (StringUtils.isNotBlank(tag)) {
+          config.setTag(tag);
+        }
+        break;
+      }
+    }
+    return config;
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/pinot/PinotThirdEyeClientConfig.java]: I recommend adding at least some javadoc stating that there can only be a single instance of the Pinot client. Or better yet, throw an exception if there's multiple.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/RCAFrameworkRunner.java [hunk 3]
```diff
@@ -139,7 +144,12 @@ public static void main(String[] args) throws Exception {
 
     ThirdEyeConfiguration thirdEyeConfig = new ThirdEyeAnomalyConfiguration();
     thirdEyeConfig.setRootDir(config.getAbsolutePath());
-    ThirdEyeCacheRegistry.initializeCaches(thirdEyeConfig);
+    String clientConfigPath = thirdEyeConfig.getClientsPath();
+    ClientConfig clientConfig = ClientConfigLoader.fromClientConfigPath(clientConfigPath);
+    if (clientConfig == null) {
+      throw new IllegalStateException("Could not create client config from path " + clientConfigPath);
+    }
+    ThirdEyeCacheRegistry.initializeCaches(thirdEyeConfig, clientConfig);
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/RCAFrameworkRunner.java]: This kind of code repeats over and over. I wonder if initializeCaches() should simply extract this information from thirdEyeConfig by itself.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java (hunk 1 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java]: Fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/pinot/PinotThirdEyeClientConfig.java (hunk 2 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/pinot/PinotThirdEyeClientConfig.java]: Added javadoc, and exception if multiple found

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/autoload/pinot/metrics/ConfigGenerator.java [hunk 4]
```diff
@@ -37,6 +38,7 @@ public static DatasetConfigDTO generateDatasetConfig(String dataset, Schema sche
     if (timeSpec.getTimeFormat().startsWith(TimeFormat.SIMPLE_DATE_FORMAT.toString())) {
       datasetConfigDTO.setTimezone(PDT_TIMEZONE);
     }
+    datasetConfigDTO.setClient(PinotThirdEyeClient.CLIENT_NAME);
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/autoload/pinot/metrics/ConfigGenerator.java]: can we change client to datasource?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/Client.java [hunk 5]
```diff
@@ -10,41 +10,34 @@
  */
 public class Client {
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/Client.java]: why is this called Client? it seems to represent clientConfig

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java [hunk 6]
```diff
@@ -4,15 +4,20 @@
...
+  public String getName();
 
-  List<String> getCollections() throws Exception;
+  public ThirdEyeResponse execute(ThirdEyeRequest request) throws Exception;
+
+  public List<String> getCollections() throws Exception;
 
   /** Clear any cached values. */
-  void clear() throws Exception;
+  public void clear() throws Exception;
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java]: public is optional in interfaces?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/pinot/PinotThirdEyeClient.java [hunk 7]
```diff
@@ -282,6 +260,19 @@ public void close() throws Exception {
     controllerClient.close();
   }
 
+  private boolean isValidProperties(Map<String, String> properties) {
+    boolean valid = true;
+    if (MapUtils.isEmpty(properties) ||
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/pinot/PinotThirdEyeClient.java]: break it into multiple lines. easier to parse and add new validation logic

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java (hunk 6 above)
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeClient.java]: Yes it is implicitly public, removed public

**Pull request merged**